File: leetcode/editor/cn/290_单词规律_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"strings"
	"testing"
)

func Test_givenNormal_when290_thenSuccess(t *testing.T) {
	assert.Equal(t, true, wordPattern("abba", "d a a d"))
	assert.Equal(t, false, wordPattern("abba", "d d d d"))
}

//leetcode submit region begin(Prohibit modification and deletion)
func wordPattern(pattern string, s string) bool {
	words := strings.Split(s, " ")
	if len(pattern) != len(words) {
		return false
	}
	ans := true
	p2w, w2p := map[uint8]string{}, map[string]uint8{}
	for i := 0; i < len(pattern); i++ {
		p, w := pattern[i], words[i]
		// 判断 map 中是否存在 key
		m1, e1 := p2w[p]
		m2, e2 := w2p[w]
		if !e1 && !e2 {
			p2w[p] = w
			w2p[w] = p
		} else if m1 != w || m2 != p {
			ans = false
			break
		}
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/290_单词规律_test.py
import unittest


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        self.assertTrue(self.solution.wordPattern("abba", "dog cat cat dog"))

    def test_givenFailCase1_when_thenSuccess(self):
        self.assertFalse(self.solution.wordPattern("abba", "dog dog dog dog"))


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words: list[str] = s.split(" ")
        if len(pattern) != len(words):
            return False
        ans: bool = True
        p2w: dict[str, str] = {}
        w2p: dict[str, str] = {}
        for i in range(len(pattern)):
            p: str = pattern[i]
            w: str = words[i]
            m1: str = p2w.get(p)
            m2: str = w2p.get(w)
            # python None
            if m1 is None and m2 is None:
                p2w[p] = w
                w2p[w] = p
            elif m1 != w or m2 != p:
                ans = False
                break
        return ans

    # leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
